### PR TITLE
[ehancement](nereids) Read lock table when analyze and generate query plan

### PR DESCRIPTION
**Pull request by Kikyou1997**
# Proposed changes

Issue Number: close #xxx

## Problem summary

So that we could avoid some concurrentmodifications to the basic structures in FE

Describe your changes.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [x] No
    - [ ] I don't know
2. Has unit tests been added:
    - [ ] Yes
    - [x] No
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [x] No
    - [ ] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [x] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [x] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 34.55 seconds
 load time: 429 seconds
 storage size: 17123343100 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221201115417_clickbench_pr_56221.html

File: fe/fe-core/src/main/java/org/apache/doris/nereids/CascadesContext.java [hunk 1]
```diff
@@ -197,4 +205,70 @@ private CascadesContext execute(Job job) {
         jobScheduler.executeJobPool(this);
         return this;
     }
+
+    public void addToTable(Table table) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/CascadesContext.java:209]:
```suggestion
    public void addTable(Table table) {
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/CascadesContext.java [hunk 2]
```diff
@@ -197,4 +205,70 @@ private CascadesContext execute(Job job) {
...
+    public void releaseTableReadLock() {
+        for (Table t : tables) {
+            t.readUnlock();
+        }
+    }
+
+    /**
+     * Find all tables to lock on.
+     */
+    public void extractTables() {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/CascadesContext.java:228]: add UT to ensure we could lock all tables when statement has view, cte, subquery in from clause and subquery in where/having clause

File: fe/fe-core/src/main/java/org/apache/doris/nereids/analyzer/NereidsAnalyzer.java [hunk 3]
```diff
@@ -54,4 +54,12 @@ public void analyze() {
         // check whether analyze result is meaningful
         new CheckAnalysisJob(cascadesContext).execute();
     }
+
+    public void analyze(boolean cte) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/analyzer/NereidsAnalyzer.java:58]: parameter is not use

File: fe/fe-core/src/main/java/org/apache/doris/nereids/jobs/batch/AnalyzeRulesJob.java [hunk 4]
```diff
@@ -68,4 +66,24 @@ public AnalyzeRulesJob(CascadesContext cascadesContext, Optional<Scope> scope) {
...
         ));
     }
+
+    @Override
+    public void execute() {
+        boolean locked = false;
+        try {
+            Job j = bottomUpBatch(ImmutableList.of(
+                    new RegisterCTE()
+            ));
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/jobs/batch/AnalyzeRulesJob.java:76]: why move rigisterCTE job?

**Comment by 924060929** [comment]
I think table read lock should span the statement lifecycle, and use the `Stack<ReentrantReadWriteLock>` save the lock order and then we can release by the reverse order. And you should try require read lock with timeout for prevent dead lock.

**Comment by Kikyou1997** [comment]
> I think table read lock should span the statement's planner lifecycle, the view and subquery maybe reentrant the same table lock. And you should use the `Stack<ReentrantReadWriteLock>` in the StatementContext to save the lock order and then we can release by the reverse order after exit the planner, catch Throwable should be considered. And you should try require read lock with timeout for prevent dead lock.

Done, thanks for these suggestions

**Review by morrySnow — APPROVED** [review]

**Pull request merged**